cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,194 @@
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(ex);
+    }
+  }
+
+  @Override
+  public RouteConfig fetch(final ProgramId serviceId) throws NotFoundException {
+    SettableFuture<RouteConfig> settableFuture = SettableFuture.create();
+    Future<RouteConfig> future = routeConfigMap.putIfAbsent(serviceId, settableFuture);
+    if (future == null) {
+      future = getAndWatchData(serviceId, settableFuture, new ZKRouteWatcher(serviceId));
+    }
+    return getConfig(serviceId, future);
+  }
+
+  private RouteConfig getConfig(ProgramId serviceId, Future<RouteConfig> future) throws NotFoundException {
+    try {
+      return future.get(ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      if (e.getCause() instanceof KeeperException.NoNodeException) {
+        throw new NotFoundException(String.format("Route Config for Service %s was not found.", serviceId));
+      }
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static String getZKPath(ProgramId serviceId) {
+    return String.format("/routestore/%s", ServiceDiscoverable.getName(serviceId));
+  }
+
+  static final Codec<RouteConfig> ROUTE_CONFIG_CODEC = new Codec<RouteConfig>() {
+
+    @Override
+    public byte[] encode(RouteConfig object) throws IOException {
+      return Bytes.toBytes(GSON.toJson(object.getRoutes()));
+    }
+
+    @Override
+    public RouteConfig decode(byte[] data) throws IOException {
+      Map<String, Integer> routes = GSON.fromJson(Bytes.toString(data), MAP_STRING_INTEGER_TYPE);
+      return new RouteConfig(routes);
+    }
+  };
+
+  private Future<RouteConfig> getAndWatchData(final ProgramId serviceId,
+                                              final SettableFuture<RouteConfig> settableFuture,
+                                              final Watcher watcher) {
+    OperationFuture<NodeData> future = zkClient.getData(getZKPath(serviceId), watcher);
+    Futures.addCallback(future, new FutureCallback<NodeData>() {
+      @Override
+      public void onSuccess(NodeData result) {
+        try {
+          RouteConfig route = ROUTE_CONFIG_CODEC.decode(result.getData());
+          settableFuture.set(route);
+          routeConfigMap.put(serviceId, settableFuture);
+        } catch (Exception ex) {
+          LOG.debug("Unable to deserialize the config for service {}. Got data {}", serviceId, result.getData());
+          // Need to remove the future from the map since later calls will continue to use this future and will think
+          // that there is an exception
+          routeConfigMap.remove(serviceId);
+          settableFuture.setException(ex);
+        }
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        routeConfigMap.remove(serviceId);
+        settableFuture.setException(t);
```

Comment:
Without caching the node not exists result, calling `getConfig` multiple times will result in a lot of calls to ZK.